### PR TITLE
#21 1.Rate limit for authenticated user and anonymous user

### PR DESCRIPTION
**Pull request by ShrijanL**
- Initiated rate limiting by DRF throttling for authenticated and anonymous user.
- Raises 429 error if predefined limit exceeds

### CODE REVIEW

File: django_generic_api/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# django_generic_api/__init__.py
```
Comment (nareshelequin) [inline django_generic_api/__init__.py:1]: missing documentation.

File: django_generic_api/utils.py [hunk 2]
```diff
@@ -142,3 +144,35 @@ def validate_char_field(value):
     "BooleanField": validate_bool_field,
     "CharField": validate_char_field,
 }
+
+
+def timeparse(period):
```
Comment (nareshelequin) [inline django_generic_api/utils.py:149]: `time_parse`, instead timeparse

File: django_generic_api/views.py [hunk 3]
```diff
@@ -25,10 +29,16 @@
...
+from .utils import (
+    ExtendedRateThrottle,
+    make_permission_str,
+    registration_token,
+    store_user_ip,
+)
 
 
 class GenericSaveAPIView(APIView):
+    throttle_classes = [UserRateThrottle]
```
Comment (nareshelequin) [inline django_generic_api/views.py:41]: try with using throttle_classes and should be same behavior.

**Review by nareshelequin — CHANGES_REQUESTED** [review]
update required changes.

File: django_generic_api/__init__.py (hunk 1 above)
Comment (ShrijanL) [reply django_generic_api/__init__.py:1]: updated documentation

File: django_generic_api/views.py (hunk 3 above)
Comment (ShrijanL) [reply django_generic_api/views.py:41]: removed throttles at class based views

File: django_generic_api/utils.py [hunk 4]
```diff
@@ -142,3 +145,17 @@ def validate_char_field(value):
...
 }
+
+
+def custom_exception_handler(exc, context):
+    response = exception_handler(exc, context)
+
+    if isinstance(exc, Throttled):
+        response = Response(
+            {
+                "error": "Too many requests. Please wait for sometime.",
```
Comment (nareshelequin) [inline django_generic_api/utils.py:156]: update message with,
Too many requests. Please try after sometime.

File: django_generic_api/utils.py [hunk 5]
```diff
@@ -142,3 +145,17 @@ def validate_char_field(value):
...
+                "code": "DGA-U003",
```
Comment (nareshelequin) [inline django_generic_api/utils.py:157]: update error code with error code documentation.

**Review by nareshelequin — CHANGES_REQUESTED** [review]
required review changes.

File: django_generic_api/utils.py (hunk 4 above)
Comment (ShrijanL) [reply django_generic_api/utils.py]: changed error message

File: django_generic_api/utils.py (hunk 5 above)
Comment (ShrijanL) [reply django_generic_api/utils.py:157]: updated in errorcode.md

**Review by nareshelequin — APPROVED** [review]

**Pull request merged**